arm9/source/utils/nandcmac.c

```diff
@@ -67,15 +72,105 @@ u32 SetupSlot0x30(char drv) {
+    AgbSaveHeader agbsave;
+    u8 upper_counter[0x1];
+    u8 lower_counter[0x1];
+    UINT br;
+    
+	if (fvx_qread(path, upper_counter, 0x34, 0x1, NULL) != FR_OK) return 1;
+	if (fvx_qread(path, lower_counter, 0x2234, 0x1, NULL) != FR_OK) return 1;
+	
+	if ((*upper_counter == 0xFF) && (*lower_counter == 0x00)); // First save initialized. Bottom slot is newer.
+	else if ((*upper_counter == 0x00) && (*lower_counter == 0xFF)) return 1; // Top slot is newer.
+	else if (*upper_counter > *lower_counter) return 1; // Compare top and bottom slots.
+	
+    if ((fvx_qread(path, &agbsave, 0x2200, 0x200, &br) != FR_OK) || (br != 0x200)) return 1;
+	
+	return ValidateAgbSaveHeader(&agbsave);
+}
+
+u32 CheckAgbSaveHeaderSram256k(const char* path) {
+    AgbSaveHeader agbsave;
+    u8 upper_counter[0x1];
+    u8 lower_counter[0x1];
+    UINT br;
+    
+	if (fvx_qread(path, upper_counter, 0x34, 0x1, NULL) != FR_OK) return 1;
+	if (fvx_qread(path, lower_counter, 0x8234, 0x1, NULL) != FR_OK) return 1;
+	
+	if ((*upper_counter == 0xFF) && (*lower_counter == 0x00)); // First save initialized. Bottom slot is newer.
+	else if ((*upper_counter == 0x00) && (*lower_counter == 0xFF)) return 1; // Top slot is newer.
+	else if (*upper_counter > *lower_counter) return 1; // Compare top and bottom slots.
+	
+    if ((fvx_qread(path, &agbsave, 0x8200, 0x200, &br) != FR_OK) || (br != 0x200)) return 1;
+	
+	return ValidateAgbSaveHeader(&agbsave);
+}
+
+u32 CheckAgbSaveHeaderFlash512k(const char* path) {
+    AgbSaveHeader agbsave;
+    u8 upper_counter[0x1];
+    u8 lower_counter[0x1];
+    UINT br;
+    
+	if (fvx_qread(path, upper_counter, 0x34, 0x1, NULL) != FR_OK) return 1;
+	if (fvx_qread(path, lower_counter, 0x10234, 0x1, NULL) != FR_OK) return 1;
+	
+	if ((*upper_counter == 0xFF) && (*lower_counter == 0x00)); // First save initialized. Bottom slot is newer.
+	else if ((*upper_counter == 0x00) && (*lower_counter == 0xFF)) return 1; // Top slot is newer.
+	else if (*upper_counter > *lower_counter) return 1; // Compare top and bottom slots.
+	
+    if ((fvx_qread(path, &agbsave, 0x10200, 0x200, &br) != FR_OK) || (br != 0x200)) return 1;
+	
+	return ValidateAgbSaveHeader(&agbsave);
+}
+
+u32 CheckAgbSaveHeaderFlash1M(const char* path) {
+    AgbSaveHeader agbsave;
+    u8 upper_counter[0x1];
+    u8 lower_counter[0x1];
+    UINT br;
+
+	if (fvx_qread(path, upper_counter, 0x34, 0x1, NULL) != FR_OK) return 1;
+	if (fvx_qread(path, lower_counter, 0x20234, 0x1, NULL) != FR_OK) return 1;
+    
+	if ((*upper_counter == 0xFF) && (*lower_counter == 0x00)); // First save initialized. Bottom slot is newer.
+	else if ((*upper_counter == 0x00) && (*lower_counter == 0xFF)) return 1; // Top slot is newer.
+	else if (*upper_counter > *lower_counter) return 1; // Compare top and bottom slots.
+	
+    if ((fvx_qread(path, &agbsave, 0x20200, 0x200, &br) != FR_OK) || (br != 0x200)) return 1;
+	
+	return ValidateAgbSaveHeader(&agbsave);
+}
```

Comment:
All the `CheckAgbSaveHeaderX` functions have too much repeated code - reading the `upper_counter` and the `if..else if block` is always the same, as is the return expression. The only things that vary are the offset for the `lower_counter` and `agbsave` read, and that information can easily be consolidated into a single structure.

There should be a single `CheckAgbSaveType` that tries to verify if it's any of the types and once it gets a match, return the type (or return a type that indicates it's not an AGB save).